### DISCUSSION TITLE
Not enough track groups

### DISCUSSION THREAD

**Discussion by spacek531** [OpenRCT2/OpenRCT2#16159] [Feature requests] [answered]
I'm working on adding two new track pieces to the flying coaster and implementing drawing functions for 8 more:

- Up60ToUp90
- Up90
- Up90ToUp60
- Down60ToDown90
- Down90
- Down90ToDown60
- MultiDimInvertedFlatToDown90QuarterLoop
- MultiDimUp90ToInvertedFlatQuarterLoop
- LeftFlyerLargeHalfLoopUp (NEW)
- RightFlyerLargeHalfLoopUp (NEW)
- RightFlyerLargeHalfLoopDown (NEW)
- LeftFlyerLargeHalfLoopDown (NEW)

The issue comes to the lack of available track piece groups:
https://github.com/OpenRCT2/OpenRCT2/blob/a396289a7a4c324f7584b75287cff92302e35657/src/openrct2/ride/Track.h#L112

There's 63 flags and a 64-bit value. There's not enough space to add enough new track piece groups that the track elements with paint code can be implemented without also including track elements that don't have new paint code, and some other issues.

For reference:

TRACK_HALF_LOOP_UNINVERTED unlocks FlyerHalfLoopUp. Used on flying rc and laydown rc.
TRACK_HALF_LOOP_INVERTED unlocks FlyerHalfLoopDown. Used on flying rc and laydown rc.
TRACK_QUARTER_LOOP_UNINVERTED unlocks MultiDimFlatToDown90QuarterLoop and MultiDimUp90ToInvertedFlatQuarterLoop. Used on multidim rc.
TRACK_QUARTER_LOOP_INVERTED unlocks and MultiDimInvertedFlatToDown90QuarterLoop and MultiDimInvertedUp90ToFlatQuarterLoop. Used on multidim rc.

FlyerLargeHalfLoopUp should receive a separate track piece group than FlyerLargeHalfLoopDown if following the game's existing setup is important. They cannot be placed in the TRACK_HALF_LOOP_* groups because there are no sprites for vekoma large half loops.

TRACK_QUARTER_LOOP_* both have one track element I do not want to implement for flyer because no track sprites exist. Specifically, TRACK_QUARTER_LOOP_UNINVERTED implements the unwanted MultiDimFlatToDown90QuarterLoop and TRACK_QUARTER_LOOP_INVERTED implements the unwanted MultiDimInvertedUp90ToFlatQuarterLoop.

I'm open to suggestions on how to solve this. Here's some that I thought of:
- do not add any elements in the 'all drawable track pieces' cheat and users must use a plugin or other low-level tools to access the pieces (no such plugin exists to my knowledge)
- keep the unwanted elements in the 'all drawable track pieces' cheat and let the users cope with the un-implemented track elements
- add a second field for track groups

**Comment by duncanspumpkin** [comment]
```
    uint64_t EnabledTrackPieces;
    uint64_t ExtraTrackPieces;
    uint64_t CoveredTrackPieces;
```

We will need to make some custom wrapper of `std::bitset` to handle this I think. `std::bitset` is too annoying to use in this case unfortunately.

**Reply by spacek531** [reply]
I looked at the api for std::bitset and it looks usable, was there a part of it you thought would be awkward?

EDIT: figured it out, it's quite annoying that most of its members are not `constexpr`. I'm currently trying out `std::array<bool,TRACK_GROUP_COUNT>`

**Reply by duncanspumpkin** [reply]
The problems with it are: you can't construct from a list of values; serialisation is rubbish; probably a few other things

**Reply by spacek531** [reply]
zeh mat pointed out that `std::array<bool,N>` is not byte-efficient like I thought it was, so I'm making effectively the same thing but with more utility.

**Comment by spacek531** [comment, marked as answer]
solved by https://github.com/OpenRCT2/OpenRCT2/pull/16172
